File: Day007_2025/Q007_GFG.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

class Solution {
  public:
    int countPairs(vector<int> &arr, int target) {
        // Complete the function
        // Q. Pair with given sum in a sorted array
        int ans = 0;
        int i = 0;
        int j = arr.size() - 1;
        while(i < j){
            int sum = arr[i] + arr[j];
            if (sum == target) {
                int countI = 0, countJ = 0;
                int valI = arr[i], valJ = arr[j];
                while (i <= j && arr[i] == valI) {
                    countI++;
                    i++;
                }
                while (i <= j && arr[j] == valJ) {
                    countJ++;
                    j--;
                }
                if (valI == valJ) {
                    ans += countI * (countI - 1) / 2;
                } 
                else {
                    ans += countI * countJ;
                }
            } 
            else if (sum < target) {
                i++;
            } 
            else {
                j--;
            }
        }
        return ans;
    }
};

//{ Driver Code Starts.
int main() {

    int t;
    cin >> t;
    cin.ignore();
    while (t--) {
        vector<int> arr;
        int target;
        string input;
        getline(cin, input);

        stringstream ss(input);
        int number;
        while (ss >> number) {
            arr.push_back(number);
        }
        cin >> target;
        cin.ignore();
        Solution obj;
        cout << obj.countPairs(arr, target) << endl;
        cout << "~\n";
    }

    return 0;
}
// } Driver Code Ends

File: Day007_2025/Q007_GFG.py
class Solution:
    def countPairs(self, arr, target):
        # Q. Pair with given sum in a sorted array
        ans = 0
        i, j = 0, len(arr) - 1

        while i < j:
            sum_ = arr[i] + arr[j]

            if sum_ == target:
                countI, countJ = 0, 0
                valI, valJ = arr[i], arr[j]

                while i <= j and arr[i] == valI:
                    countI += 1
                    i += 1
                while i <= j and arr[j] == valJ:
                    countJ += 1
                    j -= 1

                if valI == valJ:
                    ans += (countI * (countI - 1)) // 2
                else:
                    ans += countI * countJ

            elif sum_ < target:
                i += 1
            else:
                j -= 1

        return ans


# Driver code
if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        arr = list(map(int, input().strip().split()))
        target = int(input())
        obj = Solution()
        print(obj.countPairs(arr, target))
        print("~")
